File: sdk/typescript/src/http/ApiClient.ts
import axios, { AxiosInstance, AxiosRequestConfig } from 'axios';
import {
  AuthenticationError, AuthorizationError, ConsentRequiredError,
  IdempotencyConflictError, NotFoundException, OpesCareError,
  RateLimitError, ServerError, ValidationError,
} from '../errors';

const MAX_RETRIES   = 4;
const RETRY_DELAYS  = [1000, 2000, 4000, 8000]; // ms, exponential

export class ApiClient {
  private readonly http: AxiosInstance;

  constructor(baseUrl: string, private accessToken: string, timeout = 30000) {
    this.http = axios.create({
      baseURL: baseUrl.replace(/\/$/, '') + '/',
      timeout,
      headers: {
        Accept:         'application/json',
        'Content-Type': 'application/json',
        'User-Agent':   'opescare-ts-sdk/1.0',
      },
    });
  }

  async get<T = Record<string, unknown>>(path: string, params?: Record<string, unknown>): Promise<T> {
    return this.request<T>('GET', path, { params });
  }

  async post<T = Record<string, unknown>>(
    path: string,
    data?: Record<string, unknown>,
    idempotencyKey?: string
  ): Promise<T> {
    return this.request<T>('POST', path, { data }, idempotencyKey);
  }

  async delete<T = Record<string, unknown>>(path: string): Promise<T> {
    return this.request<T>('DELETE', path);
  }

  private async request<T>(
    method: string,
    path: string,
    config: AxiosRequestConfig = {},
    idempotencyKey?: string
  ): Promise<T> {
    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.accessToken}`,
    };

    if (idempotencyKey && ['POST', 'PUT', 'PATCH'].includes(method)) {
      headers['Idempotency-Key'] = idempotencyKey;
    }

    let attempt = 0;

    while (true) {
      try {
        const response = await this.http.request<T>({
          method,
          url: path.replace(/^\//, ''),
          ...config,
          headers: { ...headers, ...(config.headers ?? {}) },
        });
        return response.data;
      } catch (err: unknown) {
        if (!axios.isAxiosError(err)) throw err;

        const status      = err.response?.status ?? 0;
        const body        = err.response?.data as Record<string, unknown> ?? {};
        const errorCode   = (body.error_code ?? body.error ?? '') as string;
        const message     = (body.message ?? `HTTP ${status}`) as string;

        if (status === 429) {
          const retryAfter = parseInt(err.response?.headers['retry-after'] ?? '60', 10);
          if (attempt < MAX_RETRIES) {
            await sleep(Math.max(retryAfter * 1000, RETRY_DELAYS[attempt] ?? 8000));
            attempt++;
            continue;
          }
          throw new RateLimitError(message, retryAfter, body);
        }

        if (status >= 500 && attempt < MAX_RETRIES) {
          await sleep(RETRY_DELAYS[attempt] ?? 8000);
          attempt++;
          continue;
        }

        throw this.mapError(status, errorCode, message, body);
      }
    }
  }

  private mapError(
    status: number,
    errorCode: string,
    message: string,
    body: Record<string, unknown>
  ): OpesCareError {
    if (status === 401) return new AuthenticationError(message, status, body);
    if (status === 403 && errorCode === 'CONSENT_REQUIRED') return new ConsentRequiredError(message, status, body);
    if (status === 403) return new AuthorizationError(message, status, body);
    if (status === 404) return new NotFoundException(message, status, body);
    if (status === 409) return new IdempotencyConflictError(message, status, body);
    if (status === 422) return new ValidationError(message, status, body);
    if (status >= 500)  return new ServerError(message, status, body);
    return new OpesCareError(message, status, body);
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

File: sdk/python/src/opescare/http/api_client.py
"""
Low-level HTTP client wrapping httpx.

Handles:
  - Bearer token injection on every request
  - Idempotency-Key header on write operations
  - Exponential backoff retry for 429 / 5xx
  - Typed exception mapping
"""
from __future__ import annotations

import time
from typing import Any

import httpx

from ..exceptions import (
    AuthenticationError,
    AuthorizationError,
    ConsentRequiredError,
    IdempotencyConflictError,
    NotFoundException,
    OpesCareError,
    RateLimitError,
    ServerError,
    ValidationError,
)

_MAX_RETRIES  = 4
_RETRY_DELAYS = [1, 2, 4, 8]  # seconds, exponential
_WRITE_METHODS = {"POST", "PUT", "PATCH"}


class ApiClient:
    def __init__(self, base_url: str, access_token: str, timeout: float = 30.0) -> None:
        self._base_url     = base_url.rstrip("/")
        self._access_token = access_token
        self._timeout      = timeout
        self._http         = httpx.Client(
            base_url=self._base_url + "/",
            timeout=timeout,
            headers={
                "Accept":     "application/json",
                "User-Agent": "opescare-python-sdk/1.0",
            },
        )

    # ── Public HTTP methods ────────────────────────────────────────────────

    def get(self, path: str, params: dict[str, Any] | None = None) -> dict[str, Any]:
        return self._request("GET", path, params=params)

    def post(
        self,
        path: str,
        body: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        return self._request("POST", path, body=body, idempotency_key=idempotency_key)

    def delete(self, path: str) -> dict[str, Any]:
        return self._request("DELETE", path)

    # ── Internal ──────────────────────────────────────────────────────────

    def _request(
        self,
        method: str,
        path: str,
        params: dict[str, Any] | None = None,
        body: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        headers: dict[str, str] = {
            "Authorization": f"Bearer {self._access_token}",
        }
        if body is not None:
            headers["Content-Type"] = "application/json"
        if idempotency_key and method.upper() in _WRITE_METHODS:
            headers["Idempotency-Key"] = idempotency_key

        attempt = 0
        while True:
            try:
                response = self._http.request(
                    method,
                    path.lstrip("/"),
                    params=params,
                    json=body,
                    headers=headers,
                )
            except httpx.RequestError as exc:
                if attempt < _MAX_RETRIES:
                    time.sleep(_RETRY_DELAYS[attempt])
                    attempt += 1
                    continue
                raise ServerError(f"Request failed after {_MAX_RETRIES} retries: {exc}") from exc

            status = response.status_code

            if status < 400:
                return _decode(response)

            resp_body  = _safe_json(response)
            error_code = resp_body.get("error_code") or resp_body.get("error") or ""
            message    = resp_body.get("message") or f"HTTP {status}"

            if status == 429:
                retry_after = int(response.headers.get("Retry-After", 60))
                if attempt < _MAX_RETRIES:
                    time.sleep(max(retry_after, _RETRY_DELAYS[attempt]))
                    attempt += 1
                    continue
                raise RateLimitError(message, retry_after, resp_body)

            if status >= 500 and attempt < _MAX_RETRIES:
                time.sleep(_RETRY_DELAYS[attempt])
                attempt += 1
                continue

            raise _map_error(status, error_code, message, resp_body)

    def close(self) -> None:
        self._http.close()

    def __enter__(self) -> "ApiClient":
        return self

    def __exit__(self, *_: object) -> None:
        self.close()


def _decode(response: httpx.Response) -> dict[str, Any]:
    if not response.content:
        return {}
    try:
        return response.json()
    except Exception:
        return {"raw": response.text}


def _safe_json(response: httpx.Response) -> dict[str, Any]:
    try:
        return response.json()
    except Exception:
        return {}


def _map_error(
    status: int, error_code: str, message: str, body: dict[str, Any]
) -> OpesCareError:
    if status == 401:
        return AuthenticationError(message, status, body)
    if status == 403 and error_code == "CONSENT_REQUIRED":
        return ConsentRequiredError(message, status, body)
    if status == 403:
        return AuthorizationError(message, status, body)
    if status == 404:
        return NotFoundException(message, status, body)
    if status == 409:
        return IdempotencyConflictError(message, status, body)
    if status == 422:
        return ValidationError(message, status, body)
    if status >= 500:
        return ServerError(message, status, body)
    return OpesCareError(message, status, body)
